File: unity/Assets/Scripts/motion/IntentDispatcher.cs
// IntentDispatcher — viewer/src/motion/intentDispatcher.ts의 C# 1:1 이식.
// ESP→viewer 메시지 단일 진입점. ARCHITECTURE.md dispatch matrix 동일.
//
// 흐름 (TS와 동일):
//   IViewerConnection.OnMessage → Dispatch(msg)
//   ┌── hello/state/intent/heartbeat/error → store.ApplyEspMessage
//   ├── motion_started → store.ApplyEspMessage + motion.PlayAsync
//   ├── motion_completed → cid stale 검증 → motion.Settle({Completed, ActualMs})
//   └── motion_failed   → cid stale 검증 → motion.Settle({Failed, Reason})
//
//   IMotionController.OnEvent → store.ApplyMotionEvent

#nullable enable

using System;
using System.Collections.Generic;
using Nubjuk.Viewer.Controller;
using Nubjuk.Viewer.Store;
using Nubjuk.Viewer.Types;
using Nubjuk.Viewer.Ws;

namespace Nubjuk.Viewer.Motion
{
    public sealed class IntentDispatcher : IDisposable
    {
        private readonly IViewerConnection _conn;
        private readonly IMotionController _motion;
        private readonly ViewerStore _store;
        private readonly List<ISubscription> _subs = new();

        public IntentDispatcher(IViewerConnection conn, IMotionController motion, ViewerStore store)
        {
            _conn = conn;
            _motion = motion;
            _store = store;
            _subs.Add(_conn.OnMessage(Dispatch));
            _subs.Add(_conn.OnConnectionChange(_store.SetConnectionState));
            _subs.Add(_conn.OnProtocolError(err =>
                _store.RecordProtocolError($"[{err.Kind}] {err.Message}")));
            _subs.Add(_motion.OnEvent(_store.ApplyMotionEvent));
        }

        public void Dispose()
        {
            foreach (var s in _subs) s.Dispose();
            _subs.Clear();
        }

        private void Dispatch(EspMessage msg)
        {
            _store.ApplyEspMessage(msg);

            switch (msg)
            {
                case MotionStartedMessage started:
                    // controller가 watchdog + Started event 책임. 같은 cid 두 번은 idempotent.
                    _ = _motion.PlayAsync(started.Payload.Intent, started.Payload.ExpectedDurationMs, started.CorrelationId);
                    break;

                case MotionCompletedMessage completed:
                    {
                        var currentCid = _motion.GetCurrentCorrelationId();
                        if (currentCid != null && currentCid != completed.CorrelationId)
                        {
                            _store.RecordProtocolError(
                                $"stale motion_completed cid={completed.CorrelationId} (current={currentCid})");
                            break;
                        }
                        // ESP terminal authority — Settle이 watchdog 정리 + Completed event emit.
                        _motion.Settle(new MotionSettleResult.Completed(
                            CorrelationId: completed.CorrelationId,
                            ActualMs: completed.Payload.ActualDurationMs));
                        break;
                    }

                case MotionFailedMessage failed:
                    {
                        var currentCid = _motion.GetCurrentCorrelationId();
                        if (currentCid != null && currentCid != failed.CorrelationId)
                        {
                            _store.RecordProtocolError(
                                $"stale motion_failed cid={failed.CorrelationId} (current={currentCid})");
                            break;
                        }
                        _motion.Settle(new MotionSettleResult.Failed(
                            CorrelationId: failed.CorrelationId,
                            Reason: failed.Payload.Reason));
                        break;
                    }

                // hello/state/intent/heartbeat/error는 store가 자체 처리
                case HelloMessage:
                case StateMessage:
                case IntentMessage:
                case ErrorMessage:
                case HeartbeatMessage:
                    break;

                // 미래에 새 EspMessage subtype이 추가되어 dispatch 누락 시 silent drop 방지.
                // C# abstract record는 exhaustiveness 컴파일 검사 X — runtime 가드 필요.
                default:
                    _store.RecordProtocolError(
                        $"unhandled message type in dispatcher: {msg.GetType().Name}");
                    break;
            }
        }
    }
}

File: src/motion/intentDispatcher.ts
// IntentDispatcher — ESP→viewer 메시지 단일 진입점.
// ARCHITECTURE.md dispatch matrix + ESP-as-truth + correlation_id stale 거부.
//
// 흐름:
//   ViewerConnection.onMessage → dispatch(msg)
//   ┌── hello/state/intent/heartbeat/error → store.applyEspMessage
//   ├── motion_started   → store.applyEspMessage + motion.play
//   ├── motion_completed → cid stale 검증 → motion.settle({completed, actualMs})
//   └── motion_failed    → cid stale 검증 → motion.settle({failed, reason})
//
//   MotionController.onEvent → store.applyMotionEvent

import type { EspMessage } from "@/types/protocol";
import type { ViewerConnection, Subscription } from "@/ws/ViewerConnection";
import type { MotionController } from "@/controller/MotionController";
import type { ViewerStore } from "@/store/viewerStore";

export class IntentDispatcher {
  private subs: Subscription[] = [];

  constructor(
    private readonly conn: ViewerConnection,
    private readonly motion: MotionController,
    private readonly store: ViewerStore,
  ) {
    this.subs.push(conn.onMessage((msg) => this.dispatch(msg)));
    this.subs.push(conn.onConnectionChange((state) => store.setConnectionState(state)));
    this.subs.push(
      conn.onProtocolError((err) =>
        store.recordProtocolError(`[${err.kind}] ${err.message}`),
      ),
    );
    this.subs.push(motion.onEvent((ev) => store.applyMotionEvent(ev)));
  }

  dispose(): void {
    this.subs.forEach((s) => s.dispose());
    this.subs = [];
  }

  private dispatch(msg: EspMessage): void {
    // 모든 메시지를 store에 기록 (recent + 부분 처리)
    this.store.applyEspMessage(msg);

    switch (msg.type) {
      case "motion_started":
        // controller가 timer + started event 책임. 같은 cid 두 번은 controller가 idempotent.
        this.motion.play(
          msg.payload.intent,
          msg.payload.expected_duration_ms,
          msg.correlation_id,
        );
        break;

      case "motion_completed": {
        const currentCid = this.motion.getCurrentCorrelationId();
        if (currentCid && currentCid !== msg.correlation_id) {
          this.store.recordProtocolError(
            `stale motion_completed cid=${msg.correlation_id} (current=${currentCid})`,
          );
          break;
        }
        // ESP terminal authority — settle()이 watchdog 정리 + completed event emit.
        this.motion.settle({
          type: "completed",
          correlationId: msg.correlation_id,
          actualMs: msg.payload.actual_duration_ms,
        });
        break;
      }

      case "motion_failed": {
        const currentCid = this.motion.getCurrentCorrelationId();
        if (currentCid && currentCid !== msg.correlation_id) {
          this.store.recordProtocolError(
            `stale motion_failed cid=${msg.correlation_id} (current=${currentCid})`,
          );
          break;
        }
        this.motion.settle({
          type: "failed",
          correlationId: msg.correlation_id,
          reason: msg.payload.reason,
        });
        break;
      }

      // hello/state/intent/heartbeat/error는 store가 자체 처리
      case "hello":
      case "state":
      case "intent":
      case "error":
      case "heartbeat":
        break;
    }
  }
}
